File: crates/tracepilot-tauri-bindings/src/commands/mcp.rs
//! MCP Tauri commands (11 commands).

use crate::blocking_cmd;
use crate::config::SharedConfig;
use crate::error::CmdResult;
use crate::helpers::read_config;
use std::collections::HashMap;

// -- Server CRUD --

#[tauri::command]
#[tracing::instrument(level = "debug", skip(state), err)]
pub async fn mcp_list_servers(
    state: tauri::State<'_, SharedConfig>,
) -> CmdResult<Vec<(String, tracepilot_orchestrator::mcp::types::McpServerConfig)>> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!(tracepilot_orchestrator::mcp::config::list_servers_in(&home))
}

#[tauri::command]
#[tracing::instrument(level = "debug", skip(state), err, fields(server = %name))]
pub async fn mcp_get_server(
    state: tauri::State<'_, SharedConfig>,
    name: String,
) -> CmdResult<tracepilot_orchestrator::mcp::types::McpServerConfig> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!(tracepilot_orchestrator::mcp::config::get_server_in(
        &home, &name
    ))
}

#[tauri::command]
#[tracing::instrument(skip(state, config), err, fields(server = %name))]
pub async fn mcp_add_server(
    state: tauri::State<'_, SharedConfig>,
    name: String,
    config: tracepilot_orchestrator::mcp::types::McpServerConfig,
) -> CmdResult<()> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!(tracepilot_orchestrator::mcp::config::add_server_in(
        &home, &name, config
    ))
}

#[tauri::command]
#[tracing::instrument(skip(state, config), err, fields(server = %name))]
pub async fn mcp_update_server(
    state: tauri::State<'_, SharedConfig>,
    name: String,
    config: tracepilot_orchestrator::mcp::types::McpServerConfig,
) -> CmdResult<()> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!(tracepilot_orchestrator::mcp::config::update_server_in(
        &home, &name, config
    ))
}

#[tauri::command]
#[tracing::instrument(skip(state), err, fields(server = %name))]
pub async fn mcp_remove_server(
    state: tauri::State<'_, SharedConfig>,
    name: String,
) -> CmdResult<tracepilot_orchestrator::mcp::types::McpServerConfig> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!(tracepilot_orchestrator::mcp::config::remove_server_in(
        &home, &name
    ))
}

#[tauri::command]
#[tracing::instrument(skip(state), err, fields(server = %name))]
#[specta::specta]
pub async fn mcp_toggle_server(
    state: tauri::State<'_, SharedConfig>,
    name: String,
) -> CmdResult<bool> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!(tracepilot_orchestrator::mcp::config::toggle_server_in(
        &home, &name
    ))
}

// -- Health checks --

#[tauri::command]
#[tracing::instrument(skip(state), err)]
pub async fn mcp_check_health(
    state: tauri::State<'_, SharedConfig>,
) -> CmdResult<HashMap<String, tracepilot_orchestrator::mcp::health::McpHealthResultCached>> {
    let home = read_config(&state).copilot_home();
    let config = tokio::task::spawn_blocking(move || {
        tracepilot_orchestrator::mcp::config::load_config_in(&home)
    })
    .await??;

    let results =
        tracepilot_orchestrator::mcp::health::check_all_servers(&config.mcp_servers).await;
    Ok(results)
}

#[tauri::command]
#[tracing::instrument(skip(state), err, fields(server = %name))]
pub async fn mcp_check_server_health(
    state: tauri::State<'_, SharedConfig>,
    name: String,
) -> CmdResult<tracepilot_orchestrator::mcp::health::McpHealthResultCached> {
    let home = read_config(&state).copilot_home();
    let config = tokio::task::spawn_blocking(move || {
        tracepilot_orchestrator::mcp::config::get_server_in(&home, &name).map(|c| (name, c))
    })
    .await??;

    let result =
        tracepilot_orchestrator::mcp::health::check_single_server(&config.0, &config.1).await;
    Ok(result)
}

// -- Import --

#[tauri::command]
#[tracing::instrument(skip(path), err, fields(path_len = path.len()))]
pub async fn mcp_import_from_file(
    path: String,
) -> CmdResult<tracepilot_orchestrator::mcp::import::McpImportResult> {
    blocking_cmd!(tracepilot_orchestrator::mcp::import::import_from_file(
        std::path::Path::new(&path),
    ))
}

#[tauri::command]
#[tracing::instrument(skip(path), err, fields(%owner, %repo, git_ref = git_ref.as_deref().unwrap_or("")))]
pub async fn mcp_import_from_github(
    owner: String,
    repo: String,
    path: Option<String>,
    git_ref: Option<String>,
) -> CmdResult<tracepilot_orchestrator::mcp::import::McpImportResult> {
    blocking_cmd!(tracepilot_orchestrator::mcp::import::import_from_github(
        &owner,
        &repo,
        path.as_deref(),
        git_ref.as_deref(),
    ))
}

// -- Diff --

#[tauri::command]
#[tracing::instrument(skip(state, incoming), err, fields(incoming_count = incoming.len()))]
pub async fn mcp_compute_diff(
    state: tauri::State<'_, SharedConfig>,
    incoming: HashMap<String, tracepilot_orchestrator::mcp::types::McpServerConfig>,
) -> CmdResult<tracepilot_orchestrator::mcp::diff::McpConfigDiff> {
    let home = read_config(&state).copilot_home();
    blocking_cmd!({
        let config = tracepilot_orchestrator::mcp::config::load_config_in(&home)?;
        Ok::<_, crate::error::BindingsError>(tracepilot_orchestrator::mcp::diff::compute_diff(
            &config.mcp_servers,
            &incoming,
        ))
    })
}

File: packages/client/src/mcp.ts
/** MCP client IPC wrappers. */

import type {
  McpConfigDiff,
  McpHealthResultCached,
  McpImportResult,
  McpServerConfig,
} from "@tracepilot/types";
import { type CommandName } from "./commands.js";
import { invokePlugin, isTauri } from "./invoke.js";

async function invoke<T>(cmd: CommandName, args?: Record<string, unknown>): Promise<T> {
  if (isTauri()) {
    return invokePlugin<T>(cmd, args);
  }
  console.warn(`[TracePilot] Not in Tauri — no mock for MCP "${cmd}"`);
  throw new Error(`No mock data for MCP command: ${cmd}`);
}

// -- Server CRUD --

export async function mcpListServers(): Promise<
  [string, McpServerConfig][]
> {
  return invoke<[string, McpServerConfig][]>("mcp_list_servers");
}

export async function mcpGetServer(name: string): Promise<McpServerConfig> {
  return invoke<McpServerConfig>("mcp_get_server", { name });
}

export async function mcpAddServer(
  name: string,
  config: McpServerConfig,
): Promise<void> {
  return invoke<void>("mcp_add_server", { name, config });
}

export async function mcpUpdateServer(
  name: string,
  config: McpServerConfig,
): Promise<void> {
  return invoke<void>("mcp_update_server", { name, config });
}

export async function mcpRemoveServer(
  name: string,
): Promise<McpServerConfig> {
  return invoke<McpServerConfig>("mcp_remove_server", { name });
}

export async function mcpToggleServer(name: string): Promise<boolean> {
  return invoke<boolean>("mcp_toggle_server", { name });
}

// -- Health checks --

export async function mcpCheckHealth(): Promise<
  Record<string, McpHealthResultCached>
> {
  return invoke<Record<string, McpHealthResultCached>>("mcp_check_health");
}

export async function mcpCheckServerHealth(
  name: string,
): Promise<McpHealthResultCached> {
  return invoke<McpHealthResultCached>("mcp_check_server_health", { name });
}

// -- Import --

export async function mcpImportFromFile(
  path: string,
): Promise<McpImportResult> {
  return invoke<McpImportResult>("mcp_import_from_file", { path });
}

export async function mcpImportFromGitHub(
  owner: string,
  repo: string,
  path?: string,
  gitRef?: string,
): Promise<McpImportResult> {
  return invoke<McpImportResult>("mcp_import_from_github", {
    owner,
    repo,
    path: path ?? null,
    gitRef: gitRef ?? null,
  });
}

// -- Diff --

export async function mcpComputeDiff(
  incoming: Record<string, McpServerConfig>,
): Promise<McpConfigDiff> {
  return invoke<McpConfigDiff>("mcp_compute_diff", { incoming });
}
